simd/src/Kokkos_Simd_Scalar.hpp

```diff
@@ -0,0 +1,299 @@
+    return mask_type(m_value > other.m_value);
+  }
+  KOKKOS_FORCEINLINE_FUNCTION mask_type operator<=(simd const& other) const {
+    return mask_type(m_value <= other.m_value);
+  }
+  KOKKOS_FORCEINLINE_FUNCTION mask_type operator>=(simd const& other) const {
+    return mask_type(m_value >= other.m_value);
+  }
+  KOKKOS_FORCEINLINE_FUNCTION mask_type operator==(simd const& other) const {
+    return mask_type(m_value == other.m_value);
+  }
+  KOKKOS_FORCEINLINE_FUNCTION mask_type operator!=(simd const& other) const {
+    return mask_type(m_value != other.m_value);
+  }
+  KOKKOS_FORCEINLINE_FUNCTION void copy_from(T const* ptr,
+                                             element_aligned_tag) {
+    m_value = *ptr;
+  }
+  KOKKOS_FORCEINLINE_FUNCTION void copy_to(T* ptr, element_aligned_tag) const {
+    *ptr = m_value;
+  }
+  KOKKOS_FORCEINLINE_FUNCTION reference operator[](std::size_t) {
+    return m_value;
+  }
+  KOKKOS_FORCEINLINE_FUNCTION value_type operator[](std::size_t) const {
+    return m_value;
+  }
+};
+
+template <class T>
+[[nodiscard]] KOKKOS_FORCEINLINE_FUNCTION simd<T, simd_abi::scalar> abs(
+    simd<T, simd_abi::scalar> const& a) {
+  return simd<T, simd_abi::scalar>(std::abs(static_cast<T>(a)));
+}
+
+template <class T>
+[[nodiscard]] KOKKOS_FORCEINLINE_FUNCTION simd<T, simd_abi::scalar> sqrt(
+    simd<T, simd_abi::scalar> const& a) {
+  return simd<T, simd_abi::scalar>(std::sqrt(static_cast<T>(a)));
+}
+
+template <class T>
+KOKKOS_FORCEINLINE_FUNCTION simd<T, simd_abi::scalar> fma(
+    simd<T, simd_abi::scalar> const& x, simd<T, simd_abi::scalar> const& y,
+    simd<T, simd_abi::scalar> const& z) {
+  return simd<T, simd_abi::scalar>((static_cast<T>(x) * static_cast<T>(y)) +
+                                   static_cast<T>(z));
+}
+
+template <class T>
+KOKKOS_FORCEINLINE_FUNCTION simd<T, simd_abi::scalar> condition(
+    desul::Impl::dont_deduce_this_parameter_t<
+        simd_mask<T, simd_abi::scalar>> const& a,
+    simd<T, simd_abi::scalar> const& b, simd<T, simd_abi::scalar> const& c) {
+  return simd<T, simd_abi::scalar>(static_cast<bool>(a) ? static_cast<T>(b)
+                                                        : static_cast<T>(c));
+}
+
+template <class T, class Abi>
+[[nodiscard]] KOKKOS_FORCEINLINE_FUNCTION simd<T, Abi> copysign(
+    simd<T, Abi> const& a, simd<T, Abi> const& b) {
+  return std::copysign(static_cast<T>(a), static_cast<T>(b));
+}
+
+template <class T>
+class const_where_expression<simd_mask<T, simd_abi::scalar>,
+                             simd<T, simd_abi::scalar>> {
+ public:
+  using abi_type   = simd_abi::scalar;
+  using value_type = simd<T, abi_type>;
+  using mask_type  = simd_mask<T, abi_type>;
+
+ protected:
+  value_type& m_value;
+  mask_type const& m_mask;
+
+ public:
+  KOKKOS_FORCEINLINE_FUNCTION
+  const_where_expression(mask_type const& mask_arg, value_type const& value_arg)
+      : m_value(const_cast<value_type&>(value_arg)), m_mask(mask_arg) {}
```

Comment:
Why are you casting the `const` away?  Where do you need non-`const` access?

Reply:
This is copied from another effort to implement `std::simd`:
https://github.com/VcDevel/std-simd/blob/c69cb8f6a8627c186427b08662b05693176c73b2/experimental/bits/simd.h#L3096

The issue is that `where_expression` needs to derive from `const_where_expression`, so to avoid having a duplicate non-const reference to the SIMD value in `where_expression`, `const_where_expression` actually stores a non-const reference to the SIMD value and uses `const_cast`. It is not my favorite design but it seems to be what the "expert" settled on